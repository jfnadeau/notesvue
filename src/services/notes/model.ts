export interface Note {
    title: string
    description: string
}

export type NoteProp = { (): Note }
