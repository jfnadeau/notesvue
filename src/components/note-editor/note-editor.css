.note-editor-header,
.note-editor-body {
  padding: 1em 2em;
}

.note-editor-header {
  border-bottom: 1px solid #ddd;
}

.note-editor-title {
  margin: 0.5em 0 0;
}

.note-editor-controls {
  margin-top: 2em;
  text-align: right;
}

.note-editor-controls .secondary-button {
  margin-bottom: 0.3em;
}
