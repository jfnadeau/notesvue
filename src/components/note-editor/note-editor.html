<div>

    <input type="text" v-model="note.title">
    <textarea v-model="note.description" cols="30" rows="10"></textarea>
    <input type="button" value="Save" @click="save">

</div>