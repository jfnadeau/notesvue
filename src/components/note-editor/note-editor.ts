import { Note, NoteProp } from './../../services/notes/model';
import Vue from "vue";
import withRender from "./note-editor.html"

export const NoteEditor = withRender(Vue.extend({

    props: {
        initialNote: {
            type: Object as NoteProp
        }
    },
    data() {

        let data = {
            note: {} as Note
        };

        data.note = this.initialNote ? this.initialNote : { title: "", description: "" };
        return data;
    },
    watch: {
        initialNote: function (newInitialNote: Note) {
            this.note = { ...newInitialNote };
        }
    },
    methods: {
        save() {
            this.$emit("save", { ...this.note });
        }
    }
}));