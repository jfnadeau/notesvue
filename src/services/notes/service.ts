import { Note } from './model';

export const getNotes = () =>
    new Promise<Note[]>(
        (resolve, reject) =>
            setTimeout(() =>
                resolve(
                    [
                        { title: "Mock Note 1", description: "Mock Desc 1" },
                        { title: "Mock Note 2", description: "Mock Desc 2" },
                        { title: "Mock Note 3", description: "Mock Desc 3" }
                    ]
                )
            )
    )

export const persistNote = (note: Note) =>
    new Promise<Note[]>(
        (resolve, reject) =>
            setTimeout(() =>
                resolve()
            )
    )