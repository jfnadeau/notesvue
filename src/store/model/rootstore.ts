import { Note } from "../../services/notes/model";

export interface RootStore {
    notes: Note[];
    selectedIndex: number | null;

    hasError: boolean;
    message: string | null;
}
