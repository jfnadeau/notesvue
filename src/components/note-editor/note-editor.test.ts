import Vue from "vue";
import { createRenderer } from "vue-server-renderer";
import { shallow, Wrapper } from "vue-test-utils";

import { Note } from "../../services/notes/model";
import { NoteEditor } from "./note-editor";

describe("NoteEditor", () => {

  it("should render like snapshot", () => {

    const sut = shallow(NoteEditor);
    expect(createRenderer().renderToString(sut.vm)).resolves.toMatchSnapshot();

  });

  it("should render note content when note prop is updated", async () => {

    const newNote: Note = {
      title: "new title",
      description: "new description",
    };

    const sut = shallow(NoteEditor);
    sut.setProps({ initialNote: newNote });

    expect(createRenderer().renderToString(sut.vm)).resolves.toMatchSnapshot();
  });

  it("should emit save event when save button is clicked", () => {

    const sut = shallow(NoteEditor);
    setInputValue(sut, "input[name='title']", "Input title");
    setInputValue(sut, "textarea[name='description']", "Input desc");

    sut.find("button").trigger("click");

    expect(sut.emitted("save")[0]).toEqual(
      [{ title: "Input title", description: "Input desc" } as Note],
    );

  });

});

function setInputValue<V extends Vue>(
  wrapper: Wrapper<V>,
  selector: string,
  value: string) {

  const inputWrapper = wrapper.find(selector);
  (inputWrapper.element as HTMLInputElement).value = value;
  inputWrapper.trigger("input");
}
