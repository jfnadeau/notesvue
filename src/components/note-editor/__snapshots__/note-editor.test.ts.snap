// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteEditor should render like snapshot 1`] = `
<div class="note-editor">
    <div class="note-editor-header pure-g">
        <div class="pure-u-1-2">
            <h1 class="note-editor-title"></h1>
        </div>
        <div class="note-editor-controls pure-u-1-2">
            <button class="secondary-button pure-button">Save</button>
        </div>
    </div>
    <div class="note-editor-body">
        <form class="pure-form pure-form-stacked">
            <fieldset>
                <label for="title">Title</label>
                <input name="title" type="text" value="" class="pure-input-1">
                <label for="description">Description</label>
                <textarea name="description" rows="20" class="pure-input-1"></textarea>
            </fieldset>
        </form>
    </div>
</div>
`;

exports[`NoteEditor should render note content when note prop is updated 1`] = `
<div class="note-editor">
    <div class="note-editor-header pure-g">
        <div class="pure-u-1-2">
            <h1 class="note-editor-title">new title</h1>
        </div>
        <div class="note-editor-controls pure-u-1-2">
            <button class="secondary-button pure-button">Save</button>
        </div>
    </div>
    <div class="note-editor-body">
        <form class="pure-form pure-form-stacked">
            <fieldset>
                <label for="title">Title</label>
                <input name="title" type="text" value="new title" class="pure-input-1">
                <label for="description">Description</label>
                <textarea name="description" rows="20" class="pure-input-1">new description</textarea>
            </fieldset>
        </form>
    </div>
</div>
`;
